File: cartesianTree.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>

typedef struct _Node Node;

struct _Node {
    char key;
    int priority;
    Node* left;
    Node* right;
};


Node* createNode(char key, int priority) {
    Node* newNode = (Node*) malloc(sizeof(Node));

    if (!newNode) {
        perror("Failed to allocate memory for node creation");
        exit(EXIT_FAILURE);
    }

    newNode->key = key;
    newNode->priority = priority;
    newNode->left = NULL;
    newNode->right = NULL;

    return newNode;
}

Node* createEmptyTree() {
    return NULL;
}

int isTreeEmpty(Node* root) {
    return root == NULL;
}

Node* getLeftChild(Node* node) {
    if (!node) {
        fprintf(stderr, "Error: Cannot get left child of a NULL node.\n");
        return NULL;
    }

    return node->left;
}

Node* getRightChild(Node* node) {
    if (!node) {
        fprintf(stderr, "Error: Cannot get right child of a NULL node.\n");
        return NULL;
    }

    return node->right;
}

Node* findKey(char key, Node* root) {
    if (isTreeEmpty(root)) {
        fprintf(stderr, "Error: Empty tree.\n");
        return NULL;
    }

    Node* iter = root;
    while (iter) {
        if (iter->key == key) {
            return iter;
        }

        else {
            if (key < iter->key) {
                iter = getLeftChild(iter);
            }

            else {
                iter = getRightChild(iter);
            }
        }
    }

    return NULL;
}

Node* rotateRight(Node* root) {
    Node* newRoot = root->left;
    root->left = newRoot->right;
    newRoot->right = root;
    
    return newRoot;
}

Node* rotateLeft(Node* root) {
    Node* newRoot = root->right;
    root->right = newRoot->left;
    newRoot->left = root;
    
    return newRoot;
}

Node* BSTInsert(Node* root, Node* node) {
    if (node->priority < root->priority) {
        if (node->key < root->key) {
            node->right = root;
            return node;
        }

        else {
            node->left = root;
            return node;
        }
    }

    if (node->key < root->key) {
        if (!root->left) {
            root->left = node;
            return root;
        }

        root->left = BSTInsert(root->left, node);
        return root;
    }

    if (!root->right) {
        root->right = node;
        return root;
    }

    root->right = BSTInsert(root->right, node);
    return root;
}


Node* createBSTTree(Node* nodes[], int len) {
    Node* root = nodes[0];

    for (int i = 1; i < len; i++) {
        Node* node = nodes[i];
        root = BSTInsert(root, node);
    }

    return root;
}

Node* insertNode(Node* root, Node* node) {
    if (isTreeEmpty(root)) {
        return node;
    }

    if (node->key < root->key) {
        root->left = insertNode(root->left, node);
    }

    else if (node->key > root->key) {
        root->right = insertNode(root->right, node);
    }

    else {
        return root;
    }

    // Rotate the tree if the priority of the node is lower than the priority of the root
    if (node->priority < root->priority) {
        if (node->key < root->key) {
            return rotateRight(root);
        }

        return rotateLeft(root);
    }

    // If a rotation is not needed, return the root for the recursive calls
    return root;
}

Node* createTree(Node* nodes[], int len) {
    Node* root = nodes[0];

    for (int i = 1; i < len; i++) {
        Node* node = nodes[i];
        root = insertNode(root, node);
    }

    return root;
}

Node* deleteNode(Node* root, Node* node) {
    if (!root) {
        fprintf(stderr, "Error: Node not found.\n");
        return NULL;
    }

    // Search for the node to delete
    if (node->key < root->key) {
        root->left = deleteNode(root->left, node);
        return root;
    }
    if (node->key > root->key) {
        root->right = deleteNode(root->right, node);
        return root;
    }

    // Node to delete is found
    if (!root->left && !root->right) {
        // Case 1: Node is a leaf
        free(root);
        return NULL;
    }
    
    else if (!root->left) {
        // Case 2: Node has only a right child
        Node* temp = root->right;
        free(root);
        return temp;
    }
    
    else if (!root->right) {
        // Case 3: Node has only a left child
        Node* temp = root->left;
        free(root);
        return temp;
    }

    // Case 4: Node has two children
    // Rotate with the child that has the lowest priority
    if (root->left->priority < root->right->priority) {
        root = rotateRight(root);
        root->right = deleteNode(root->right, node);
    } else {
        root = rotateLeft(root);
        root->left = deleteNode(root->left, node);
    }

    return root;
}

void showNode(Node* node) {
    if (node) {
        printf("key: %c, priority: %d\n", node->key, node->priority);
    }
}

void showTree(Node* root) {
    if (!root) {
        return;
    }

    showNode(root);
    
    if (root->left) {
        printf("-left of %c%d\n", root->key, root->priority);
        showTree(root->left);
    }

    if (root->right) {
        printf("-right of %c%d\n", root->key, root->priority);
        showTree(root->right);
    }
}


int main() {
    // Create nodes
    Node* nodeA = createNode('A', 5);
    Node* nodeB = createNode('B', 3);
    Node* nodeC = createNode('C', 8);
    Node* nodeD = createNode('D', 2);
    Node* nodeE = createNode('E', 6);
    Node* nodeF = createNode('F', 7);
    Node* nodeG = createNode('G', 9);
    Node* nodeH = createNode('H', 1);
    Node* nodeI = createNode('I', 10);
    Node* nodeJ = createNode('J', 12);

    Node* nodes[10] = {nodeA, nodeB, nodeC, nodeD, nodeE, nodeF, nodeG, nodeH, nodeI, nodeJ};
    int len = 10;

    // Create empty tree
    Node* tree = createEmptyTree();
    
    // Insertion
    tree = createTree(nodes, len);
    
    // Deletion
    tree = deleteNode(tree, nodeA);
    tree = deleteNode(tree, nodeJ);
    tree = deleteNode(tree, nodeI);
    showTree(tree);
    
    return 0;
}

File: CartesianTree.py
import binarytree
import random
import time
import numpy as np
import matplotlib.pyplot as plt

# --------------------------------------------------------------------------- #
# CODE FOR EXERCISE 1      (STRUCTURE AND PROPERTIES)                         #
# Class and visualisation functions, for the Cartesian Tree data structure    #
# --------------------------------------------------------------------------- #

class Node:
    def __init__(self, key, priority, left=None, right=None):
        self.key = key
        self.priority = priority
        self.left = left
        self.right = right
    
    def __str__(self):
        return f'Node({self.key}, {self.priority})'
    
    def copy(self):
        return Node(self.key, self.priority, self.left, self.right)

def nodes_from_string(nodes_str):
    nodes = []
    for node_str in nodes_str.split(','):
        key, priority = node_str.strip('()').split(' : ')
        nodes.append(Node(key, int(priority)))
    return nodes

def print_nodes(nodes):
    for node in nodes:
        print(node)
        
def copy_nodeslist(nodes):
    return [node.copy() for node in nodes]

def convert_to_printable(root, highlighted_node=None):
    if root is None:
        return None
    if highlighted_node is not None and root.key == highlighted_node.key:
        return binarytree.Node('**(' + root.key + ', ' + str(root.priority) + ')**', convert_to_printable(root.left, highlighted_node), convert_to_printable(root.right, highlighted_node))
    return binarytree.Node('(' + root.key + ', ' + str(root.priority) + ')', convert_to_printable(root.left, highlighted_node), convert_to_printable(root.right, highlighted_node))

def print_tree(root, highlighted_node=None):
    print(convert_to_printable(root, highlighted_node=highlighted_node))
    
def get_depth(root):
    if root is None:
        return 0
    return 1 + max(get_depth(root.left), get_depth(root.right))

def get_depth_of_node(root, node):
    if root == node:
        return get_depth(root)
    if root is None:
        return 0
    if root.key == node.key:
        return 1
    if node.key < root.key:
        return 1 + get_depth_of_node(root.left, node)
    return 1 + get_depth_of_node(root.right, node)


# ------------------------------------------------- #
# CODE FOR EXERCISE 2      (FIND NODE)              #
# ------------------------------------------------- #

def find_node(root, key):
    if root is None:
        return None
    if root.key == key:
        return root
    if key < root.key:
        return find_node(root.left, key)
    return find_node(root.right, key)



# ------------------------------------------------- #
# CODE FOR EXERCISE 3      (INSERT NODE)            #
# ------------------------------------------------- #

def BST_insert(root, node, verbose=False):
    if node.key < root.key:
        if root.left is None:
            root.left = node
            if verbose:
                print(f"Added node to left of root ({root.key}.left = {node.key})")
            return root
        if verbose:
            print(f"\tGoing one layer down on the left from {root.key} for {node.key}")
        root.left = BST_insert(root.left, node, verbose=verbose)
        return root
    if root.right is None:
        root.right = node
        if verbose:
            print(f"Added node to right of root ({root.key}.right = {node.key})")
        return root
    if verbose:
        print(f"\tGoing one layer down on the right from {root.key} for {node.key}")
    root.right = BST_insert(root.right, node, verbose=verbose)
    return root

def create_tree_BST(nodes, verbose=False):
    root = nodes[0]
    for node in nodes[1:]:
        root = BST_insert(root, node, verbose=verbose)
    return root


def rotate_tree_right(root):
    new_root = root.left
    root.left = new_root.right
    new_root.right = root
    return new_root

def rotate_tree_left(root):
    new_root = root.right
    root.right = new_root.left
    new_root.left = root
    return new_root


def insert_node(root, node, verbose=False):
    if root is None:
        if verbose:
            print(f"Inserting {node.key}")
        return node
    
    # Continue the tree until we find a leaf
    if node.key < root.key:
        if verbose:
            if root.left is not None:
                print(f"Going one layer down on the left from {root.key} to {root.left.key} for {node.key}")
            else:
                print(f"Going one layer down to add {node.key} to the left of {root.key}")
        root.left = insert_node(root.left, node, verbose=verbose)
    elif node.key > root.key:
        if verbose:
            if root.right is not None:
                print(f"Going one layer down on the right from {root.key} to {root.right.key} for {node.key}")
            else:
                print(f"Going one layer down to add {node.key} to the right of {root.key}")
        root.right = insert_node(root.right, node, verbose=verbose)
    else:
        if verbose:
            print(f"Node {node.key} already exists in the tree")
        return root
    
    
    # Rotate the tree if the priority of the node is lower than the priority of the root
    if node.priority < root.priority:
        if node.key < root.key:
            if verbose:
                print(f"Rotating right at {root.key} for {node.key}")
            return rotate_tree_right(root)
        if verbose:
            print(f"Rotating left at {root.key} for {node.key}")
        return rotate_tree_left(root)
    
    # If a rotation is not needed, return the root for the recursive calls
    return root
        
    
def create_tree(nodes, verbose=False):
    root = nodes[0]
    for node in nodes[1:]:
        root = insert_node(root, node, verbose=verbose)
    return root


# ------------------------------------------------- #
# CODE FOR EXERCISE 4      (REMOVE NODE)            #
# ------------------------------------------------- #


def remove_node(root, node, verbose=False):
    if root is None:
        if verbose:
            print(f"Node {node.key} not found in the tree")
        return None
    
    # Search for the node to remove
    if node.key < root.key:
        if verbose:
            print(f"-SEARCHING {node.key}- Going one layer down on the left from {root.key} for {node.key}")
        root.left = remove_node(root.left, node, verbose=verbose)
        return root
    if node.key > root.key:
        if verbose:
            print(f"-SEARCHING {node.key}- Going one layer down on the right from {root.key} for {node.key}")
        root.right = remove_node(root.right, node, verbose=verbose)
        return root
    
    # Case 1: Node is a leaf
    if root.left is None and root.right is None:
        if verbose:
            print(f"{node.key} is a leaf node, removing it")
        return None
    
    # Case 2: Node has one child
    elif root.left is None:
        # Case 2.1: Node has a right child
        if verbose:
            print(f"{node.key} has only a right child, removing it and replacing it with the right child")
        return root.right
    
    elif root.right is None:
        # Case 2.2: Node has a left child
        if verbose:
            print(f"{node.key} has only a left child, removing it and replacing it with the left child")
        return root.left
    
    # Case 3: Node has two children
    # Rotate with the child that has the lower priority
    if root.left.priority < root.right.priority:
        if verbose:
            print(f"{node.key} has two children, left child has lower priority, rotating right")
        root = rotate_tree_right(root)
        if verbose:
            print(f"Rotated tree, new root is {root.key}, continuing with remove_node({root.right.key}, {node.key})")
            print_tree(root)
        root.right = remove_node(root.right, node, verbose=verbose)
    else:
        if verbose:
            print(f"{node.key} has two children, right child has lower priority, rotating left")
        root = rotate_tree_left(root)
        if verbose:
            print(f"Rotated tree, new root is {root.key}, continuing with remove_node({root.left.key}, {node.key})")
            print_tree(root)
        root.left = remove_node(root.left, node, verbose=verbose)
    
    return root

def remove_key(root, key, verbose=False):
    return remove_node(root, find_node(root, key), verbose=verbose)



# ------------------------------------------------- #
# CODE FOR EXERCISE 5      (RANDOM TREES)           #
# ------------------------------------------------- #


ALPHABET = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'

def create_random_tree(n):
    key_size = len(str(n)) + 1
    priority_range = (1, n*10)
    nodes = []
    for i in range(n):
        key = ''.join(random.choices(ALPHABET, k=key_size))
        priority = random.randint(*priority_range)
        nodes.append(Node(key, priority))
    return create_tree(nodes)

def create_random_node(n):
    key_size = len(str(n)) + 1
    priority_range = (1, n*10)
    key = ''.join(random.choices(ALPHABET, k=key_size))
    priority = random.randint(*priority_range)
    return Node(key, priority)

def create_random_nodes_linspace(n, max_priority=-1):
    if max_priority == -1:
        max_priority = n * 10
    key_size = len(str(n)) + 1
    priorities = np.linspace(1, max_priority, n, dtype=int)
    nodes = []
    for i in range(n):
        key = ''.join(random.choices(ALPHABET, k=key_size))
        priority = priorities[i]
        nodes.append(Node(key, priority))
    return nodes




# ------------------------------------------------- #
# CODE FOR EXERCISE 6      (THEORETICAL ASPECT)     #
# ------------------------------------------------- #

def root_to_nodelist(root):
    if root is None:
        return []
    return [root] + root_to_nodelist(root.left) + root_to_nodelist(root.right)

def get_keylist(nodelist):
    return [node.key for node in nodelist]

def get_prioritylist(nodelist):
    return [node.priority for node in nodelist]

def get_kth_node(root, k):
    nodelist = root_to_nodelist(root)
    keylist = get_keylist(nodelist)
    sorted_keys = sorted(keylist)
    kth_key = sorted_keys[k-1]
    return find_node(root, kth_key)

def get_kth_node_depth(root, k):
    return get_depth_of_node(root, get_kth_node(root, k))